src/typescript/sdk/src/markets/utils.ts

```diff
@@ -322,3 +328,111 @@ export function calculateTvlGrowth(periodicStateTracker1D: Types["PeriodicStateT
   // (b * c) / (a * d)
   return b.mul(c).div(a.mul(d)).toString();
 }
+
+type ReservesAndBondingCurveState = Flatten<
+  AtLeastOne<{
+    inBondingCurve: boolean;
+    lpCoinSupply: bigint;
+  }> & {
+    clammVirtualReserves: Types["Reserves"];
+    cpammRealReserves: Types["Reserves"];
+  }
+>;
+
+/**
+ * Calculates the circulating supply based on the given market state.
+ *
+ * The logic for calculation is taken directly from the Move smart contract.
+ * @see assign_supply_minuend_reserves_ref_mut in `emojicoin_dot_fun.move`
+ */
+export const calculateCirculatingSupply = ({
+  clammVirtualReserves,
+  cpammRealReserves,
+  ...args
+}: ReservesAndBondingCurveState) =>
+  isInBondingCurve(args)
+    ? BASE_VIRTUAL_CEILING - clammVirtualReserves.base
+    : EMOJICOIN_SUPPLY - cpammRealReserves.base;
+
+/**
+ * *NOTE*: If you already have a market's state, call {@link calculateCirculatingSupply} directly.
+ *
+ * Fetches the circulating supply of a market by looking at its on-chain state.
+ *
+ * Uses the Aptos fullnode; be mindful of rate-limiting.
+ *
+ * @param emojis the input {@link SymbolEmoji}s that form the market symbol.
+ * @param ledgerVersion an optional ledger version number to specify the view function should use.
+ * @returns the market's circulating supply if the market exists, `undefined` otherwise.
+ */
+export const fetchCirculatingSupply = async (
+  emojis: SymbolEmoji[],
+  ledgerVersion?: AnyNumberString
+): Promise<bigint | undefined> =>
+  MarketView.view({
+    aptos: getAptosClient(),
+    marketAddress: getMarketAddress(emojis),
+    options: ledgerVersion
+      ? {
+          ledgerVersion: BigInt(ledgerVersion),
+        }
+      : {},
+  })
+    .then(toMarketView)
+    .then(calculateCirculatingSupply)
+    .catch(() => undefined);
+
+/**
+ * Calculates the real reserves of a market based on the given market state.
+ *
+ * The logic for calculating the real reserves of a market in the bonding curve is taken directly
+ * from the Move smart contract.
+ *
+ * Note that while the market is in the bonding curve, the current market price (the current
+ * position on the bonding curve) is represented by the clamm's virtual reserves, *not* the clamm's
+ * real reserves calculated here.
+ *
+ * Post bonding curve, the price is directly equal to the cpamm's real reserves.
+ *
+ * For an in depth explanation of the math and behavior behind the bonding curve:
+ * @see {@link https://github.com/econia-labs/emojicoin-dot-fun/blob/main/doc/blackpaper/emojicoin-dot-fun-blackpaper.pdf}
+ */
+export const calculateRealReserves = ({
```

Comment:
- [ ] Use the same test I mentioned above, but just calculate real reserves minus ciruclating supply and ensure it adds to supply cap